corehq/form_processor/backends/couch/processor.py

```diff
@@ -147,39 +147,45 @@ def get_cases_from_forms(case_db, xforms):
         return touched_cases
 
     @staticmethod
-    def hard_rebuild_case(domain, case_id, detail, save=True):
-        try:
-            case = CommCareCase.get(case_id)
-            assert case.domain == domain
-            found = True
-        except CaseNotFound:
+    def hard_rebuild_case(domain, case_id, detail, save=True, lock=True):
+        case, lock_obj = FormProcessorCouch.get_case_with_lock(case_id, lock=lock)
+        found = bool(case)
+        if not found:
             case = CommCareCase()
             case.case_id = case_id
             case.domain = domain
-            found = False
+            if lock:
+                lock_obj = CommCareCase.get_obj_lock_by_id(case_id)
 
-        forms = FormProcessorCouch.get_case_forms(case_id)
-        filtered_forms = [f for f in forms if f.is_normal]
-        sorted_forms = sorted(filtered_forms, key=lambda f: f.received_on)
-
-        actions = _get_actions_from_forms(domain, sorted_forms, case_id)
-
-        if not found and case.domain is None:
-            case.domain = domain
-
-        rebuild_case_from_actions(case, actions)
-        # todo: should this move to case.rebuild?
-        if not case.xform_ids:
-            if not found:
-                return None
-            # there were no more forms. 'delete' the case
-            case.doc_type = 'CommCareCase-Deleted'
-
-        # add a "rebuild" action
-        case.actions.append(_rebuild_action())
-        if save:
-            case.save()
-        return case
+        assert case.domain == domain
+        try:
+            if not found and lock_obj:
+                # the lock was only aquired if we found a case
+                acquire_lock(lock_obj, degrade_gracefully=False)
```

Comment:
Ah, now I see what you're doing. I think it would be clearer if you moved this line up to line 159.

Reply:
d6d76db